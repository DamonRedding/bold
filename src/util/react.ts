import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
