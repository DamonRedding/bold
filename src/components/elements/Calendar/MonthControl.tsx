import React from 'react'

import { getUserLocale } from '../../../util/locale'
import { capitalize } from '../../../util/string'
import { HFlow } from '../../layout/Flow/HFlow'
import { Button } from '../Button'
import { Icon } from '../Icon'

export interface MonthControlProps {
  visibleDate: Date
  onChange(newDate: Date): void
  renderMonth?(date: Date): React.ReactNode
}

export const MonthControl = (props: MonthControlProps) => {
  const { visibleDate, onChange, renderMonth } = props

  const handleNext = () => {
    const next = new Date(visibleDate)
    next.setMonth(visibleDate.getMonth() + 1)
    return onChange(next)
  }

  const handlePrev = () => {
    const prev = new Date(visibleDate)
    prev.setMonth(prev.getMonth() - 1)
    return onChange(prev)
  }

  return (
    <HFlow alignItems='center' hSpacing={0.5}>
      <Button title='Previous month' size='small' skin='ghost' onClick={handlePrev} tabIndex={-1}>
        <Icon icon='angleLeft' />
      </Button>
      {renderMonth(visibleDate)}
      <Button title='Next month' size='small' skin='ghost' onClick={handleNext} tabIndex={-1}>
        <Icon icon='angleRight' />
      </Button>
    </HFlow>
  )
}

MonthControl.defaultProps = {
  renderMonth: date => {
    const formatter = new Intl.DateTimeFormat(getUserLocale(), { month: 'short' })
    return capitalize(formatter.format(date))
  },
} as Partial<MonthControlProps>
