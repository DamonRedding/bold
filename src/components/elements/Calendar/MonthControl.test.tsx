import { mount, shallow } from 'enzyme'
import * as React from 'react'

import { withTheme } from '../../../test'

import { MonthControl } from './MonthControl'

it('should render correctly', () => {
    expect(shallow(
        <MonthControl visibleDate={new Date('2018-10-26')} onChange={jest.fn()} />
    )).toMatchSnapshot()
})

it('should call onChange function with next month when clicked on next button', () => {
    const change = jest.fn()
    const wrapper = mount(withTheme(
        <MonthControl visibleDate={new Date('2018-10-26')} onChange={change} />
    ))
    expect(change).not.toHaveBeenCalled()
    wrapper.find('button').last().simulate('click')
    expect(change).toHaveBeenCalledWith(new Date('2018-11-26'))
})

it('should call onChange function with previous month when clicked on prev button', () => {
    const change = jest.fn()
    const wrapper = mount(withTheme(
        <MonthControl visibleDate={new Date('2018-10-26')} onChange={change} />
    ))
    expect(change).not.toHaveBeenCalled()
    wrapper.find('button').first().simulate('click')
    expect(change).toHaveBeenCalledWith(new Date('2018-09-26'))
})
