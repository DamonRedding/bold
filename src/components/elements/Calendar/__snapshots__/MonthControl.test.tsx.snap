// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<HFlow
  alignItems="center"
  hSpacing={0.5}
>
  <Component
    kind="normal"
    onClick={[Function]}
    size="small"
    skin="ghost"
    tabIndex={-1}
  >
    <Component
      icon="angleLeft"
      size={1.5}
    />
  </Component>
  Oct
  <Component
    kind="normal"
    onClick={[Function]}
    size="small"
    skin="ghost"
    tabIndex={-1}
  >
    <Component
      icon="angleRight"
      size={1.5}
    />
  </Component>
</HFlow>
`;
